File: Python/44. Wildcard Matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        i = j = 0
        last_match = 0
        star_index = -1

        while i < len(s):
            print(s[i], p[j], star_index, last_match)
            if j < len(p) and (p[j] == "?" or p[j] == s[i]):
                i += 1
                j += 1
                
            elif j < len(p) and  p[j] == "*":
                last_match = i
                star_index = j
                j += 1
            
            elif star_index != -1:
                j = star_index + 1
                i = last_match + 1
                last_match += 1
            else:
                return False
            
        while j < len(p) and p[j] == "*":
            j += 1
            
        return j == len(p)
    

File: Javascript/44. Wildcard Matching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    let i = 0
    let j = 0

    let lastMatch = 0
    let star = -1

    while (i < s.length) {
        if (j < p.length && (s[i] == p[j] || p[j] == "?")) {
            i++
            j++
        } else if (j < p.length && p[j] == "*") {
            last_match = i
            star = j
            j++
        } else if (star != -1) {
            j = star + 1
            i = last_match + 1
            last_match++
        } else {
            return false
        }
    }

    while (j < p.length && p[j] == "*") {
        j++
    }

    return j == p.length
};